include/RAJA/pattern/teams/teams_core.hpp

```diff
@@ -0,0 +1,369 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   RAJA header file containing user interface for RAJA::Teams
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_pattern_teams_core_HPP
+#define RAJA_pattern_teams_core_HPP
+
+#include "RAJA/config.hpp"
+#include "RAJA/internal/get_platform.hpp"
+#include "RAJA/policy/cuda/policy.hpp"
+#include "RAJA/policy/loop/policy.hpp"
+#include "RAJA/policy/openmp/policy.hpp"
+#include "RAJA/util/macros.hpp"
+#include "RAJA/util/plugins.hpp"
+#include "RAJA/util/types.hpp"
+#include "RAJA/util/StaticLayout.hpp"
+#include "camp/camp.hpp"
+#include "camp/concepts.hpp"
+#include "camp/tuple.hpp"
+
+
+#if defined(__CUDA_ARCH__)
+#define TEAM_SHARED __shared__
+#else
+#define TEAM_SHARED
+#endif
+
+namespace RAJA
+{
+
+template<typename DataType, camp::idx_t... Sizes>
+struct TeamSharedArray
```

Comment:
Folks, here is a different way to allocate shared memory. If I try to make the array a member variable, I get a compile error. However; it looks like we can initialize shared memory in the constructor since the lifespan is the same as the CUDA block. If we go down this route this adds the cost of storing a pointer.